### ISSUE TITLE
Are data loaders shared across requests?

### ISSUE THREAD

**Issue by corstian** [graphql-dotnet/graphql-dotnet#1178]
I have seen behaviour which can be explained as follows:

1. Request 1 requests a data loader and creates a new one. This data loader uses an object instance scoped to this request.
2. Request 2 also requests a data loader and gets the one already created.
3. Request 1 finishes the request.
4. Request 2 throws an exception because the object instance from request 1 has been disposed.

I could work around this issue by supplying `Activity.Current.Id` (which I've been told is unique for each request) as part of the key identifying the data loader.

Now first I suspected my issue was registering the `IDataLoaderContextAccessor` as singleton, but I see the same is done [in the `AddDataLoader` extension method in the server package](https://github.com/graphql-dotnet/server/blob/e7d889e96a3ec7713f8aefa3d3f89b492bc99710/src/Core/GraphQLBuilderExtensions.cs).

Therefore my question: are data loaders shared across (HTTP) requests?

**Comment by johnrutherford** [comment]
DataLoaders should not be shared across requests. How are you accessing the `DataLoaderContext`?

**Comment by corstian** [comment]
Thanks! ~It happens via constructor injection whereas I have an extension method on the `IDataLoaderContextAccessor`.~

Nevermind, I just realized my problem is most certainly related to a type (creating a DB connection) which was registered as singleton, but should have been scoped, and therefore resulting in a deadlock when multiple queries were executed.

Slightly unrelated to the original issue, but is there a built in shortcut for accessing scoped services? My current method to access these services is through the `ContextProvider` type which I think I have copied from somewhere around here.

What it essentially does is to retrieve the `IServiceProvider` instance bound to the current `HttpContext` through the `IHttpContextAccessor`.
